Client, Server Socket.io 설정 및 예제코드 작성

junghyunbak opened the pull request:
## 링크(관련 이슈)

#40 

<br>

## 개요

Client, Server Socket.io 설정 및 예제 코드 작성

<br>

## 내용

### Client

- 소켓 연결 시 Absolute URL `http://localhost:3000/room` 대신 Relative URL `/room` 를 사용하기 위해 `vite.config.ts` 의 proxy 를 설정 [(개발 노트)](https://www.notion.so/delicious-guys/Frontend-Dev-Server-Proxy-with-Vite-af438cc2b29f4eeaa32635c363ef5573)
- socket 객체와, 생명 주기 관리를 위한 함수들을 모아 useSocket 커스텀훅으로 관리 [(개발 노트)](https://www.notion.so/delicious-guys/React-Socket-io-9029c4b88d9945e4b4d67f115154a6bb)
- 서버와 소켓 연결이 재대로 되었는지 확인하기 위해 `src/pages/RoomPage` 를 생성하여 예제 코드를 작성

\+ React.StrictMode 를 제거

### Server

- `room` 네임스페이스로 소켓 접속이 가능한 WebSocketGateway 를 생성 [(개발 노트)](https://www.notion.so/delicious-guys/Nest-js-Socket-io-fbcb12837f1d4277a3c574404301611c) 

<br>

## 비고

- 서버 쪽 작업을 하다 보니 서버의 package.json 의 engines 설정이 되어있지 않더라고요. 해당 pr 에 추가로 적용할 지 논의해주시면 적용하도록 하겠습니다.
- 소켓 이벤트명 네이밍 컨벤션이 정해졌었나요? 정해지지 않았다면 논의가 필요할 것 같습니다.

<br>

## 리뷰어한테 할 말

- 서버, 클라이언트 Socket.io 연동 작업을 따로 할 경우 테스트하기 어려워 하나의 issue 로 묶었는데, pr 이 조금 거대해졌습니다. 조금 보기 불편하셔도 양해 부탁드립니다.

junghyunbak committed:
chore: React.StrictMode 제거

이유 : 사용 시 두 번 렌더링되기 때문에 소켓 관련 테스트에 방해가 됨

junghyunbak committed:
chore: nest.js 소켓 사용을 위한 관련 패키지 설치

junghyunbak committed:
feat: nest.js socket.io 연동 (기본 세팅)

junghyunbak committed:
chore: socket.io-client 패키지 설치

junghyunbak committed:
chore: 개발서버 proxy 설정 (Base domain url 사용을 위함)

junghyunbak committed:
feat: 소켓 생명주기를 컨트롤하기 위한 useSocket 훅 생성

junghyunbak committed:
feat: 소켓 연결을 테스트하기 위한 room 페이지 임시 생성

One-armed-boy commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+
+  const connectSocket = (): Promise<void> => {
+    return new Promise((resolve, reject) => {
+      if (socket.current !== null) {
+        resolve();
+        // resolve or reject 함수를 호출해도
+        // 코드는 계속해서 실행되기 때문에 return
+        return;
+      }
+
+      socket.current = io('/room');
+
+      socket.current.on('connect', () => {
+        resolve();
+      });
+
+      socket.current.on('connect_error', () => {
+        reject();
+      });
+    });
+  };
```
별로 중요한 건 아니라서 꼭 반영은 안하셔도 됩니다.
socket.current 가 계속 반복되는데, connectSocket 코드 블록 안에 먼저 const socket = socket.current 처럼 선언을 먼저해두고 가져다써도 될 것 같아요

One-armed-boy commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+import { MutableRefObject, useRef } from 'react';
+
+export const useSocket = (): [MutableRefObject<Socket | null>, () => void, () => void] => {
+  const socket = useRef<Socket | null>(null);
```
코드를 쭉 읽다보니 갑자기 생각나는건데 소켓을 useState 가 아닌 useRef로 사용하신 특별한 이유가 있나요? 지금 로직도 문제는 없어보이는데 useState를 사용해도 비슷하게 코드가 짜일 것 같아서, 혹시 어떤 근거로 둘 중 useRef 를 선택하셨는지 궁금하네요.

junghyunbak replied:
소켓 객체를 상태로 사용하면, 소켓 연결이 이루어졌을 때 컴포넌트가 다시 렌더링 되면서 구현이 매우 복잡해집니다.

아래는 구현이 복잡해진다는 것을 뒷받침하는 간단한 근거입니다.

```typescript
const [socket, setSocket] = useState<Socket | null>(null):

useEffect(() => {
  setSocket(io('/room')); // 이 부분에서 렌더링이 다시 발생

  console.log('실행되지 않음');
}, []):
```

또, 소켓 객체가 직접적으로 UI 에 영향을 주지 않는다고 생각했습니다. (연결 여부에 따른 화면 렌더링은 `ready` 라는 상태가 결정합니다.)
=> 때문에 상태로 사용되어 컴포넌트가 리렌더링 되게 할 필요가 없다.
=> useRef 사용하면 훨씬 깔끔하게 코딩이 가능하다.

One-armed-boy replied:
넵, 답변 감사합니다. 
정현님이 useRef를 사용하신 특별한 이유가 있나 궁금했는데, 해소되었습니다. 감사합니다.😄

One-armed-boy commented on client/vite.config.ts:
```diff
@@ -19,4 +19,11 @@ export default defineConfig({
+  server: {
+    proxy: {
+      '/socket.io': {
+        target: 'http://localhost:3000',
```
이 부분은 개발 환경, 릴리즈 환경, 프로덕션 환경에 따라 유동적으로 사용이 가능하도록 로직이 짜여지는게 이후에 소요가 적을 것 같습니다. (ex. 환경변수를 사용하여 코드 실행환경에 따라 바뀌어야하는 URL의 하드코딩을 피한다.)

junghyunbak replied:
`vite.config.js` 의 설정은 개발서버만의 옵션이라 릴리즈, 프로덕션 환경을 고려하지 않아도 됩니다. (물론 백앤드 개발환경의 포트번호를 환경변수에 저장하여 관리하는 것은 필요하다고 생각합니다.)

### 실제 배포시 시나리오는 다음과 같습니다.

배포 시에는 개발서버 proxy 가 적용되지 않으니,

1. Relative URL 로 소켓연결 요청 (`/room/socket.io/*` => `https://chobab.site/room/socket.io/*`)
2. chobab.site 도메인 443번 포트의 요청을 NGINX 가 받아 nest.js 서버로 reverse-proxy

One-armed-boy replied:
결국 저 옵션은, dev 모드일 때만 유효한 옵션이라 localhost로 넣어놔도 문제가 없다는거군요? 감사합니다.😀

One-armed-boy commented on server/src/app.module.ts:
```diff
@@ -4,6 +4,7 @@ import { ConfigModule, ConfigService } from '@nestjs/config';
 import { MongooseModule } from '@nestjs/mongoose';
 import { RoomModule } from '@room/room.module';
 import * as Joi from 'joi';
+import { EventsGateway } from './socket/socket.gateway';
```
절대경로 설정 부탁드립니다!

junghyunbak replied:
넵 추가하도록 하겠습니다!

YuriKwon commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+import { io, Socket } from 'socket.io-client';
+import { MutableRefObject, useRef } from 'react';
+
+export const useSocket = (): [MutableRefObject<Socket | null>, () => void, () => void] => {
```
이 부분이 자동 타입 추론이 안될라나요!?
인라인으로 길게 적혀있으니까 뭔가 없애거나 따로 타입으로 빼고싶네요,,
 
개인적인 의견이므로 반영 안하셔도 됩니다ㅎㅎ

YuriKwon replied:
connetSocket 함수는 타입이 
`(): Promise<void>`인데 `() => void` 로 써도 되는건가요??
몰라서 여쭤봅니다,,

YuriKwon commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+import { MutableRefObject, useRef } from 'react';
+
+export const useSocket = (): [MutableRefObject<Socket | null>, () => void, () => void] => {
+  const socket = useRef<Socket | null>(null);
```
socket을 useRef로 선언해서 사용할 때랑, 
`let socket: Socket | null = null;`
이렇게 지역변수로 선언해서 사용할 때랑 무슨 차이점이 있나요?

junghyunbak replied:
~useState 로 소켓 객체를 관리하려다 구현이 복잡해져서 찾게 된 대안이 useRef 였는데 지역 변수와 차이가 없네요.~

~그리고 useRef 는 DOM 요소 접근을 위해 사용되는 훅이라고 하네요. useRef 를 제가 남용한 것 같습니다.~

~위에 @One-armed-boy 창명님과 나눈 대화가 부끄러워지네요 🥵~

One-armed-boy replied:
저도 엘리먼트 참조할 때 쓴다고 알고 있었는데 정현님 코드보고 '저렇게도 쓸 수 있구나~' 했는데 아닌가보군요?

유리님 말씀보고 드는 생각이, 원래 재렌더링이 발생하는 환경에서는 useRef가 계속 동일한 포인터를 참조해서 의미가 있는데, 저렇게 훅으로 따로 떼서 클로저로 활용할 때는 전혀 의미가 없어보이기도 하네요. 

하나 배워갑니당.

YuriKwon replied:
오호,, 저도 잘 몰라서 질문한거라 부끄러워하실 필요 없습니다ㅎㅎㅎ
저도 배워가요~~~

YuriKwon commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+
+  const connectSocket = (): Promise<void> => {
+    return new Promise((resolve, reject) => {
+      if (socket.current !== null) {
```
궁금한 점인데,
`socket.current`로 쓰고 계신데 여기서 `.current`를 하는 이유가 ref 때문인거 맞죠??

뭔가 거슬려서 저 current를 없애고 싶어요.. 
대강 이런식으로...?
```
const socketRef = useRef<Socket | null>(null);
const socket = socketRef.current;
```

junghyunbak committed:
refactor: socket 폴더 절대경로 지정

YuriKwon commented on client/src/pages/RoomPage/index.tsx:
```diff
@@ -0,0 +1,63 @@
+
+  const [socket, connectSocket, disconnectSocket] = useSocket();
+
+  const loading = async () => {
```
함수명이 동사로 시작하면 좋을것같아요!
load~, init~...

YuriKwon commented on client/src/pages/RoomPage/index.tsx:
```diff
@@ -0,0 +1,63 @@
+  }, []);
+
+  useEffect(() => {
+    if (socket.current === null) {
```
이 부분을 `!socket.current` 이렇게 바꿔도 동일하게 동작할까요?

junghyunbak replied:
네 똑같이 작동할겁니다. 일부러 명시적으로 작성하려고 `=== null` 로 작성했었습니다.

```javascript
if (!(socketRef.current instanceof Socket)) {
  // ...
}
```

현재는 좀 더 코드를 읽기 편하도록 위와 같이 `instanceof` 로 체크하도록 변경했습니다. 의견 주세요.

YuriKwon replied:
좋은것같아요~~

YuriKwon commented on client/src/pages/RoomPage/index.tsx:
```diff
@@ -0,0 +1,63 @@
+    socket.current.emit('clientToServer', 'message');
+  };
+
+  return !ready ? (
+    <div>loading...</div>
+  ) : (
+    <div>
+      room page
+      <button type="button" onClick={sendMessage}>
+        message send
+      </button>
+    </div>
+  );
+}
```
삼항연산자 앞에 !를 쓰는게 개인적으로 굉장히 낯선데,,
이렇게 쓰면 !연산이 필요 없지 않을까여..?

```js
return ready ? (
    <div>
      room page
      <button type="button" onClick={sendMessage}>
        message send
      </button>
    </div>
  ) : (
    <div>loading...</div>
  );
}
```

junghyunbak replied:
저는 삼항 연산자 조건문에 이보다 더 한 짓을 한 코드를 많이 봐서 그런가 충분히 자연스러워 보인다고 생각합니다,,

그리고 로딩화면 다음으로 메인화면이 등장하는 순서를 생각했을 때, loading 페이지 컴포넌트가 위쪽에 선언되었으면 좋겠다고 생각되어 `!ready` 형태로 작성했습니다.

YuriKwon replied:
화면 흐름을 생각하면 그렇네요 ㅎㅎ

YuriKwon commented on client/src/pages/RoomPage/index.tsx:
```diff
@@ -0,0 +1,63 @@
+import { useSocket } from '@hooks/useSocket';
+
+function RoomPage() {
+  const [ready, setReady] = useState<boolean>(false);
```
사소한거라 큰 의미는 없을 수도 있는데, 
ready보다 명확한 이름이면 좀 더 좋을것같아요~!

boolean 값이니까 isReady나, isCompleteLoading 같은 느낌으로..

YuriKwon commented on client/src/hooks/useSocket.ts:
```diff
@@ -0,0 +1,39 @@
+export const useSocket = (): [MutableRefObject<Socket | null>, () => void, () => void] => {
+  const socket = useRef<Socket | null>(null);
+
+  const connectSocket = (): Promise<void> => {
```
Promise, async await를 자유자재로 쓰고 계시네요!!! 
제가 비동기 쪽이 좀 약해서 정현님 코드를 보고 많이 배워갑니다,, 👍🏻

CodeDiary18 commented on client/src/Router.tsx:
```diff
@@ -12,6 +13,7 @@ function Router() {
         <Routes>
           <Route path="/" element={<HomePage />} />
           <Route path="/init-room" element={<InitRoomPage />} />
+          <Route path="/room/:id" element={<RoomPage />} />
```
사소하긴 한데 현재 모임방 식별자?를 id말고 code로 얘기가 계속되고 있고 백엔드에서도 code 키워드를 활용해 변수명을 사용하고 있습니다. 그래서 프론트도 `:id`를 `:code` 혹은 `:roomCode`로 변경하면 어떤지 의견을 남겨봅니다!

CodeDiary18 reviewed (commented):
다른 분들이 코멘트 남겨주신 거에 어느 정도 동의해서 리뷰 코멘트 하나를 제외하고 이모지로 공감하는 의견에 표시해두었습니다!
제가 프론트 쪽에서 상태 관리(useRef, ...) 같은 부분과 백엔드 쪽에서 nest에서 socket을 사용하는 방법에 대해 제대로 알지 못하다 보니 코드와 PR에 남긴 설명 만으로는 이해가 되지 않는 부분이 조금씩 있었습니다.  
이러한 부분을 보완하고자 개발 노트를 작성하자라고 얘기가 되었던 것 같은데 노션에 들어가보니 이번 PR과 관련해서 개발 노트가 작성되지 않은 것 같더라구요..  
예를 들어 @nestjs/websockets라는 패키지에서 사용하는 WebSocketGateway, SubscribeMessage, ... 등 이런 부분에 대해서 사실 유추는 할 수 있지만 기존에 사용하지 않았던 것이다 보니 제대로 이해하지 못할 수 있을 것 같습니다. 저희가 딱 업무 분담이 아니고 이어서 작업해야 하는 부분이 생길 수 있다 보니 어디 사이트를 참고 했다거나 개발에 적용한 점에 대해서 간단히라도 요약을 남겨 놓으면 리뷰나 나중에 이와 관련해 다른 사람이 작업을 진행할 때 도움이 될 것 같다는 의견을 남겨봅니다😀

junghyunbak committed:
refactor: useSocket 의 disconnectSocket 함수 제거

- useRef 컴포넌트 생명주기에 종속적이므로, 컴포넌트가 unmount 될 때 socket 객체는 자동 소멸됨.

junghyunbak committed:
refactor: review 반영

- boolean 상태 변수명 변경 ready -> isReady
- 함수명 prefix 동사로 변경 (initService)
- useRef 변수명 변경 socket -> socketRef
  - 반복되는 current 참조 제거
- route parameter 명 변경

junghyunbak committed:
feat: component umount 시 socket disconnect

- 이전 커밋에서 삭제했던 함수이지만 다시 생성한 이유
  - watch 모드에서 수정이 반영되었을 때 socket 객체 계속 생성되는 문제 발견
  - 컴포넌트 생명주기에 따라 자동 소멸되지만 명시적으로 disconnect 하기 위함

junghyunbak commented:
@CodeDiary18 넵 주말동안 개발노트 작성해서 PR 메세지에 링크해 두겠습니다.

@YuriKwon, @CodeDiary18 리뷰해주신 것들 중 반영한 부분은 코멘트에 🚀 이모지 달아두었습니다.

junghyunbak committed:
Merge branch 'dev' into setting/#40

junghyunbak committed:
chore: 사용되지 않는 package 제거

junghyunbak committed:
fix: 생략 된 @nestjs/websockets lifecycle hook 추가

junghyunbak commented:
@One-armed-boy @YuriKwon @CodeDiary18 

개발노트 모두 작성해서 메세지에 링크 걸어 두었습니다.

11/25 마클에서 이야기 했던대로 실시간성 구현에 소켓을 사용할지는 월요일날 다시 논의를 해봐야 할 것 같아서 주말에 머지할 순 없을 것 같지만, 어떻게 될 진 모르니 개발노트 확인 겸 해서 읽어보시고 감정 달아주세요
